src/Lean/Linter/Basic.lean

```diff
@@ -76,50 +76,71 @@ def unusedVariables : Linter := fun stx => do
 
   -- determine unused variables
   for (id, ⟨decl?, uses⟩) in vars.toList do
+    -- process declaration
     let some decl := decl?
       | continue
     let declStx := skipDeclIdIfPresent decl.stx
     let some range := declStx.getRange?
       | continue
     let some localDecl := decl.info.lctx.find? id
       | continue
-    if !stxRange.contains range.start || localDecl.userName.hasMacroScopes then
+    if !cmdStxRange.contains range.start || localDecl.userName.hasMacroScopes then
       continue
 
+    -- check if variable is used
+    if !uses.isEmpty || tacticFVarUses.contains id || decl.aliases.any (match · with | .fvar id => tacticFVarUses.contains id | _ => false) then
+        continue
+
+    -- check linter options
     let opts := decl.ci.options
     if !getLinterUnusedVariables opts then
       continue
 
-    let mut ignoredPatternFns := #[
+    -- collect ignore functions
+    let mut ignoreFns := #[
       isTopLevelDecl constDecls,
       matchesUnusedPattern,
       isVariable,
       isInStructure,
       isInInductive,
       isInCtorOrStructBinder,
-      isInConstantOrAxiom,
+      isInOpaqueOrAxiom,
       isInDefWithForeignDefinition,
       isInDepArrow
     ]
     if !getLinterUnusedVariablesFunArgs opts then
-      ignoredPatternFns := ignoredPatternFns.append #[
+      ignoreFns := ignoreFns.append #[
         isInLetDeclaration,
         isInDeclarationSignature,
         isInFun
       ]
     if !getLinterUnusedVariablesPatternVars opts then
-      ignoredPatternFns := ignoredPatternFns.append #[
+      ignoreFns := ignoreFns.append #[
         isPatternVar
       ]
 
-    let some stack := findSyntaxStack? stx declStx
-      | continue
-    if ignoredPatternFns.any (· declStx stack) then
+    -- evaluate ignore functions on original syntax
+    if let some stack := findSyntaxStack? cmdStx declStx then
+      if ignoreFns.any (· declStx stack) then
+        continue
+    else
+      continue
+
+    -- evaluate ignore functions on macro expansion outputs
+    if ← infoTrees.anyM fun tree => do
+      if let some macroExpansions ← collectMacroExpansions? range tree then
+        return macroExpansions.reverse.any fun expansion =>
+          if let some stack := findSyntaxStack? expansion.output declStx then
+            ignoreFns.any (· declStx stack)
+          else
+            false
+      else
+        return false
+    then
       continue
```

Comment:
Something like this might be a little clearer?
```suggestion
    for tree in infoTrees do
      if let some macroExpansions ← collectMacroExpansions? range tree then
        for expansion in macroExpansions.reverse do
          if let some stack := findSyntaxStack? expansion.output declStx then
            if ignoreFns.any (· declStx stack) then
              continue
```

Reply:
The `reverse` probably deserves a short comment

Reply:
Won't the `continue` target the innermost `for` loop instead of the one going through the found variables?

Reply:
Ahh, you're right of course.